File: 560.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        prefix_sums = defaultdict(int)
        prefix_sums[0] = 1

        prefix_sum = 0
        cnt = 0

        for num in nums:
            prefix_sum += num
            cnt += prefix_sums[prefix_sum - k]
            prefix_sums[prefix_sum] += 1

        return cnt
        

File: 560.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        map<int, int> freq;
        freq[0] = 1;
        int cnt = 0;
        int prefix = 0;
        
        for (const int &num : nums){
            prefix += num;

            if(freq[prefix - k]) cnt += freq[prefix - k];

            freq[prefix]++;
        }
        return cnt;
    }
};
